rclpy/src/rclpy/events_executor/events_executor.cpp

```diff
@@ -0,0 +1,959 @@
+  if (
+    RCL_RET_OK !=
+    rcl_service_set_on_new_request_callback(
+      rcl_ptr, RclEventCallbackTrampoline, rcl_callback_manager_.MakeCallback(rcl_ptr, cb, with)))
+  {
+    throw std::runtime_error(
+      std::string("Failed to set the on new request callback for service: ") +
+      rcl_get_error_string().str);
+  }
+}
+
+void EventsExecutor::HandleRemovedService(py::handle service)
+{
+  py::handle handle = service.attr("handle");
+  const rcl_service_t * rcl_ptr = py::cast<Service>(handle).rcl_ptr();
+  if (RCL_RET_OK != rcl_service_set_on_new_request_callback(rcl_ptr, nullptr, nullptr)) {
+    throw std::runtime_error(
+      std::string("Failed to clear the on new request callback for service: ") +
+      rcl_get_error_string().str);
+  }
+  rcl_callback_manager_.RemoveCallback(rcl_ptr);
+}
+
+void EventsExecutor::HandleServiceReady(py::handle service, size_t number_of_events)
+{
+  ran_user_ = true;
+  py::gil_scoped_acquire gil_acquire;
+
+  // Largely based on rclpy.Executor._take_service() and _execute_service().
+  // https://github.com/ros2/rclpy/blob/06d78fb28a6d61ede793201ae75474f3e5432b47/rclpy/rclpy/executors.py#L386-L397
+  const py::handle handle = service.attr("handle");
+  const py::object service_take_request = handle.attr("service_take_request");
+  const py::handle srv_type = service.attr("srv_type");
+  const py::handle req_type = srv_type.attr("Request");
+  const py::handle res_type = srv_type.attr("Response");
+  const py::handle callback = service.attr("callback");
+  const py::object send_response = service.attr("send_response");
+
+  for (size_t i = 0; i < number_of_events; ++i) {
+    py::object maybe_request_and_header = service_take_request(req_type);
+    if (!maybe_request_and_header.is_none()) {
+      py::tuple request_and_header = py::cast<py::tuple>(maybe_request_and_header);
+      py::handle request = request_and_header[0];
+      py::handle header = request_and_header[1];
+      if (!request.is_none()) {
+        py::object response;
+        try {
+          response = callback(request, res_type());
+        } catch (const py::error_already_set & e) {
+          HandleCallbackExceptionInNodeEntity(e, service, "services");
+          throw;
+        }
+        send_response(response, header);
+      }
+    }
+  }
+}
+
+void EventsExecutor::HandleAddedWaitable(py::handle waitable)
+{
+  // The Waitable API is too abstract for us to work with directly; it only exposes APIs for dealing
+  // with wait sets, and all of the rcl callback API requires knowing exactly what kinds of rcl
+  // objects you're working with.  We'll try to figure out what kind of stuff is hiding behind the
+  // abstraction by having the Waitable add itself to a wait set, then take stock of what all ended
+  // up there.  We'll also have to hope that no Waitable implementations ever change their component
+  // entities over their lifetimes.
```

Comment:
Hm, the rclcpp layer explicitly forces the waitables, to implement events handling. I don't know much about the rclpy waitable internals, but the approach taken here, would break the rclcpp waitables, so we should check this... A good candidate to have a look at should be the action interfaces.

Reply:
There are test cases in `test_events_executor.py` specifically for actions, so I know that rclpy `Waitable` type works, unless you think there's something specific missing from that test coverage.

Can you clarify what it is you're looking for here?  Are you wanting me to modify the `Waitable` interface to better support callbacks directly?  That's probably a bit more ambitious and invasive than I was hoping to be, but I can look into it if you think it's necessary.

Reply:
The expire mechanism of the action server is strange in the way, that the timer callback does not expire the goals. Instead this happens if the timer is expired / ready in the waitable mechanism. We had problems in that area in the rclcpp implementation. Therefore I would check if this still works.

Reply:
test_events_executor.py has coverage for this, unless I'm misunderstanding what you mean.
https://github.com/bmartin427/rclpy/blob/events_executor/rclpy/test/test_events_executor.py#L643-L650